boruta/1-click-boruta/script.whizzml

```diff
@@ -0,0 +1,386 @@
+;; This will eventually be packed into a library
+
+(define IMPORTANCE_INC 1)
+(define IMPORTANCE_DEC -1)
+
+;; choosable-objective-ids
+;;
+;; List of IDs of the fields in the dataset that can be chosen as objective
+;; field.
+;;
+;; Inputs:
+;;  fields: (map) Fields structure
+;; Output: (list) list of field IDs
+(define (choosable-objective-ids fields)
+  (let (field-val (lambda (fid k) (get-in fields [fid k]))
+        objective-types ["categorical" "numeric"]
+        pref? (lambda (k) (field-val k "preferred"))
+        pred? (lambda (k) (member? (field-val k "optype") objective-types)))
+    (filter (lambda (x) (and (pref? x) (pred? x))) (keys fields))))
+
+
+;; validate-objective-name
+;;
+;; Validates that the argument is a valid objective name in the reference
+;; dataset.
+;;
+;; Inputs:
+;;  objective-name: (string) Name of the objective field
+;;  dataset: (map) dataset resource information
+;;
+;; Output: (string) objective-name
+(define (validate-objective-name objective-name dataset)
+  (let (fields (get dataset "fields" {})
+        objective-ids (choosable-objective-ids fields)
+        objective-names (map (lambda (x) (get-in fields [x "name"] false))
+                        objective-ids))
+    (if (not (member? objective-name objective-names))
+      (raise {"message" (str "Failed to find the objective name in the dataset"
+                             " choosable fields.")
+              "code" 101})
+      objective-name)))
+
+
+;; boruta
+;;
+;; Marks fields as important, unimportant or neutral according to their
+;; importance compared to shadow fields (copy fields using random values)
+;;
+;; Inputs:
+;;   dataset-id: (dataset-id) ID of the origin dataset
+;;   objective: (string) Name of the objective field
+;;   max-runs: (integer) Maximum number of iterations
+;;   min-gain: (number) Increment to define the interval of minimum gain
+;;                      around the maximum importance of shadow fields
+;;
+;; Output: (map) Map of important fields and unimportant fields
+
+(define (boruta dataset-id objective max-runs min-gain)
+  (let (ds (fetch dataset-id)
+        ;; original fields are stored for later use
+        ori-fields (get ds "fields" {})
+        field-ids (keys ori-fields)
+        f-names (map (lambda (x) (get-in ori-fields [x "name"] false))
+                     field-ids)
+        init-rank (reduce (lambda (x y) (assoc x (get-in ori-fields
+                                                         [y "name"]
+                                                         false) 0))
+                          {}
```

Comment:
it's more idiomatic to cut the lambda this way:

```
(reduce (lambda (x y)
          (assoc x (ori-fields [y "name"] false) 0))
        {}
        field-ids)
```

although this is even shorter with iterate:

```
(iterate (r {} id field-ids)
  (assoc r (ori-fields [id "name"] false) 0))
```

Also, you want to use `false` as a key in the result when an id
doesn't exist instead of throwing and error? (just double-checking)